### PR TITLE
移出model和dataload

### PR DESCRIPTION
**Pull request by xiangyuli-bio**
1.baselint_datamodule.py 添加缓存和移出分段逻辑 合并TDCdataload的功能
2.dg_datamodule.py 作为基类 添加bins和classify属性
3.morgan_chembert_datmodule.py 继承baseline_datamodule.py的类
  实现同时加载morgan和chembert的功能 合并了TDCdataload_double的功能
4.base_model.py 完善一些细节 加入了CLM:loss和predict
5.morgan_model.py 实现了原本DrugAttentionmodel的功能
6.morgan_chembert_model,py 继承morganmodel
  实现了DrugAttnetionmodel_Double的功能 用于morgan
  chembert同时加载的模型
7.train.py 对于morgan_model
  需要传入drug_shape由于morgan还是chembert的不同，加入Ensemble_Learn参数来选择分类头个数

### CODE REVIEW

File: src/datamodule/morgan_chembert_datamodule.py [hunk 1]
```diff
@@ -2,56 +2,181 @@
...
+        if config.model_architecture in (
+            "MorganChembertAttention",
+        ):
+            self.attention = True
+        else:
+            self.attention = False
+
+        self.drug_featurizer_two = self.drug_featurizer[1]
+        self.drug_featurizer = self.drug_featurizer[0]
+    def prepare_data(self):
```
Comment (fireinice) [inline src/datamodule/morgan_chembert_datamodule.py:51]: 这里没有行分隔

File: src/datamodule/morgan_chembert_datamodule.py [hunk 2]
```diff
@@ -2,56 +2,181 @@
...
+        super(MorganChembertDataModule, self).prepare_data()
+        self.prepare_featurizer(self.drug_featurizer_two,self.all_drugs)
+
+    def setup(self, stage: str):
+        self.setup_featurizer(self.target_featurizer, self.all_targets)
```
Comment (fireinice) [inline src/datamodule/morgan_chembert_datamodule.py:56]: 这里有比较多的代码重复，考虑单独写一个函数

File: src/models/morgan_chembert_model.py [hunk 3]
```diff
@@ -1,67 +1,80 @@
...
-        drug, target, label = train_batch  # target is (D + N_pool)
-        pred = self.forward(drug, target,False)
-        loss = self.loss_fct(pred, label.to(torch.float32))
-        self.log("val/loss", loss)
-        result = {"loss": loss, "preds": pred, "target": label}
-        self.validation_step_outputs.append(result)
-        return result
+        out_embedding = weight_one * out_embedding_one + weight_two * out_embedding_two
+        x = self.mlp(out_embedding)
+        if self.loss_type == 'OR' and self.Ensemble_Learn:
```
Comment (fireinice) [inline src/models/morgan_chembert_model.py]: 这个变量为什么起了个这么奇怪的名字

File: src/models/morgan_chembert_model.py [hunk 4]
```diff
@@ -1,67 +1,80 @@
...
+            predict = [classifier(x) for classifier in self.predict_layer]
```
Comment (fireinice) [inline src/models/morgan_chembert_model.py]: 这部分代码应该在父类中实现一致的逻辑，这里继承复用就可以了

File: src/models/morgan_model.py [hunk 5]
```diff
@@ -13,50 +17,157 @@ def __init__(
         target_dim=1024,
         latent_dim=1024,
         classify=True,
-        num_classes=52,
-        loss_type="OR",
+        num_classes=24,
```
Comment (fireinice) [inline src/models/morgan_model.py]: 这里有的是24,有的是52,建议都统一成2

**Review by fireinice — CHANGES_REQUESTED** [review]
有一些代码可以单独写成一函数方便继承复用

File: src/datamodule/baseline_datamodule.py [hunk 6]
```diff
@@ -1,64 +1,214 @@
...
-            device=device,
-            seed=config.replicate,
-            batch_size=config.batch_size,
-            shuffle=config.shuffle,
-            num_workers=config.num_workers,
-            classify=config.classify,
-            bins=config.bins,
-        )
+    def calculate_weights(self, label_dict, dataset):
+        arr = []
```
Comment (fireinice) [inline src/datamodule/baseline_datamodule.py]: 这里的weights是基于每个item的，不是基于类的，如果有90万的条目，需要一个90万长度的list

**Review by fireinice — CHANGES_REQUESTED** [review]
看一下评审意见

File: src/datamodule/morgan_chembert_datamodule.py [hunk 7]
```diff
@@ -2,56 +2,181 @@
...
+        if stage == "test" or stage is None:
+            self.test_data = BinaryDataset_Double(
+                self.df_test[self._drug_column],
+                self.df_test[self._target_column],
+                self.df_test[self._label_column],
+                self.drug_featurizer,
+                self.drug_featurizer_two,
+                self.target_featurizer,)
+
+    def calculate_weights(self, label_dict, dataset):
```
Comment (fireinice) [inline src/datamodule/morgan_chembert_datamodule.py]: 可以参考这个实现：
    def calculate_weights(self, label_dict, dataset):
        arr = []
        for _, _, label in dataset:
            count = label_dict[label.item()]
            if count == 0:
                weight = 0
            else:
                weight = len(dataset) / count
            arr.append(weight)
        return arr

**Review by fireinice — CHANGES_REQUESTED** [review]
